File: Enigma.Net/Enigma/Roll.cs
using System;
using System.Collections.Generic;
using System.Data;

namespace Enigma
{
    public class Roll
    {
        private byte Position { get; set; }     //This is the actual position of this roll starting at 0
        private byte[] Transitions { get; }     //This is the wiring of the roll: if Transitions[0] = 0x04 the value 0x00 will be mapped to 0x04
        private List<int> TurnOverIndices { get; }  //While rolling after each char encryption the next roll will also rotate, if these indices contain Position
        private byte[] ReTransitions { get; }   //Reverted transitionlist for decryption

        public Roll(byte[] transitions, List<int> turnOverIndices)
        {
            Transitions = transitions;
            TurnOverIndices = turnOverIndices;
            Position = 0;

            ReTransitions = new byte[256];
            for (int i = 0; i < 256; i++)
                ReTransitions[Transitions[i]] = (byte)i;
        }

        public void CheckInput()
        {
            if (Transitions.Length != 256)
                throw new ArgumentOutOfRangeException("Wrong Transition length ");

            for (int i = 0; i < 256; i++)
            {
                bool found = false;
                for (int j = 0; j < 256; j++)
                    if (Transitions[j] == i)
                    {
                        found = true;
                        break;
                    }

                if(!found)
                    throw new ConstraintException("Transitions not 1-1 complete");
            }

            if (TurnOverIndices.Count != 53)
                throw new ArgumentOutOfRangeException("Wrong TurnOverIndices length ");

            TurnOverIndices.Sort();
            for(int i = 0; i < TurnOverIndices.Count - 1; i++)
                if(TurnOverIndices[i] == TurnOverIndices[i+1])
                    throw new ConstraintException("Turnoverindizes has doubles");
        }

        public byte Enc(byte input)
        {
            return Transitions[(byte)(input + Position)];
        }

        public byte Dec(byte input)
        {
            return (byte)(ReTransitions[input] - Position);
        }

        public bool RollOn()
        {
            ++Position;
            return TurnOverIndices.Contains(Position);
        }
    }
}

File: Enigma.Java/src/main/Roll.java
package main;

import java.util.Collections;
import java.util.List;


public class Roll
 {
    private int position;          //This is the actual position of this roll starting at 0
    private byte[] transitions;     //This is the wiring of the roll: if Transitions[0] = 0x04 the value 0x00 will be mapped to 0x04
    private List<Integer> turnOverIndices;  //While rolling after each char encryption the next roll will also rotate, if these indices contain Position 
    private byte[] reTransitions;   //Reverted transitionlist for decryption
    private static final int  UNSIGNED_BYTE_MAX = 256;	
    private static final int  UNSIGNED_BYTE_OFFSET = 128;	//to calculate with unsigned bytes, this offset moves the range to [0,255]
    

    public Roll(byte[] transitions, List<Integer> turnOverIndices) {
        this.transitions = transitions;
        this.turnOverIndices = turnOverIndices;
        position = 0;

        reTransitions = new byte[UNSIGNED_BYTE_MAX];
        for (int i = -UNSIGNED_BYTE_OFFSET; i < UNSIGNED_BYTE_OFFSET; i++)
            reTransitions[transitions[i + UNSIGNED_BYTE_OFFSET] + UNSIGNED_BYTE_OFFSET] = (byte) i;
    }

    public void CheckInput(int transitionCount) throws Exception {
        if (transitions.length != UNSIGNED_BYTE_MAX)
            throw new Exception("Wrong Transition length ");

        for (int i = -UNSIGNED_BYTE_OFFSET; i < UNSIGNED_BYTE_OFFSET; i++) {
            boolean found = false;
            for (int j = 0; j < UNSIGNED_BYTE_MAX; j++)
            {
                if (transitions[j] == i)
                {
                    found = true;   
                    break;
                }
            }
            if (!found)
                throw new Exception("Transitions not 1-1 complete. Problem at " + i);
        }

        if (turnOverIndices.size() != transitionCount)
            throw new Exception("Wrong TurnOverIndices length ");

        Collections.sort(turnOverIndices);

        for (int i = 0; i < turnOverIndices.size() - 1; i++)
            if (turnOverIndices.get(i) == turnOverIndices.get(i + 1))
                throw new Exception("Turnoverindizes has doubles");
    }

    public byte Encrypt(byte input) {
        return transitions[((input + position + UNSIGNED_BYTE_OFFSET)%UNSIGNED_BYTE_MAX)];
    }

    public byte Decrypt(byte input) {
        return (byte) (reTransitions[input + UNSIGNED_BYTE_OFFSET] - position);
    }

    public boolean RollOn() {
        position = (position + 1) % UNSIGNED_BYTE_MAX;
        return turnOverIndices.contains(position);
    }
}
